### PR TITLE
Add cgo and lapack64 functions for performing a QR and LQ solve from …

### PR DESCRIPTION
**Pull request by btracey**
…the factorization

### CODE REVIEW

**Comment by btracey** [comment]
The test for Dormlq is disabled because I believe there is a bug in the c interface. Line 56 of http://www.netlib.org/lapack/explore-html/d1/de6/lapacke__dormlq__work_8c_source.html  tests that the stride of a is greater than m. However, the number of columns of A depends on the Side, so the test should compare against m or n. See our check on lines 34 to 44 of  lapack/native/Dorml2

**Comment by kortschak** [comment]
I agree.

**Comment by btracey** [comment]
OpenBLAS bug submitted at https://github.com/xianyi/OpenBLAS/issues/615#issuecomment-129617183

**Comment by btracey** [comment]
Lapack bug submitted but it is under moderation.

**Comment by btracey** [comment]
PTAL @kortschak

File: cgo/lapack.go [hunk 1]
```diff
@@ -327,3 +327,114 @@ func (impl Implementation) Dgetrs(trans blas.Transpose, n, nrhs int, a []float64
...
+// A is of size k×n. This uses a blocked algorithm.
+//
+// Work is temporary storage, and lwork specifies the usable memory length.
+// At minimum, lwork >= m if side == blas.Left and lwork >= n if side == blas.Right,
+// and this function will panic otherwise.
+// Dormlq uses a block algorithm, but the block size is limited
+// by the temporary space available. If lwork == -1, instead of performing Dormlq,
+// the optimal work length will be stored into work[0].
+//
+// Tau contains the householder scales and must have length at least k, and
```
Comment (kortschak) [inline cgo/lapack.go]: tau

File: cgo/lapack.go [hunk 2]
```diff
@@ -327,3 +327,114 @@ func (impl Implementation) Dgetrs(trans blas.Transpose, n, nrhs int, a []float64
...
+		if lwork < m {
+			panic(badWork)
+		}
+	}
+
+	clapack.Dormqr(side, trans, m, n, k, a, lda, tau, c, ldc)
+}
+
+// Dtrtrs solves a triangular system of the form A * X = B or A^T * X = B. Dtrtrs
+// returns whether A is singular. If A is singular, no solve is performed.
```
Comment (kortschak) [inline cgo/lapack.go]: This suggests that a return of true means singular. This is not the case.

File: lapack64/lapack64.go [hunk 3]
```diff
@@ -162,3 +162,53 @@ func Getrf(a blas64.General, ipiv []int) bool {
...
+// A is of size k×n. This uses a blocked algorithm.
+//
+// Work is temporary storage, and lwork specifies the usable memory length.
+// At minimum, lwork >= m if side == blas.Left and lwork >= n if side == blas.Right,
+// and this function will panic otherwise.
+// Ormlq uses a block algorithm, but the block size is limited
+// by the temporary space available. If lwork == -1, instead of performing Ormlq,
+// the optimal work length will be stored into work[0].
+//
+// Tau contains the householder scales and must have length at least k, and
```
Comment (kortschak) [inline lapack64/lapack64.go]: tau

File: lapack64/lapack64.go [hunk 4]
```diff
@@ -162,3 +162,53 @@ func Getrf(a blas64.General, ipiv []int) bool {
...
+// and this function will panic otherwise.
+// Ormqr uses a block algorithm, but the block size is limited
+// by the temporary space available. If lwork == -1, instead of performing Ormqr,
+// the optimal work length will be stored into work[0].
+func Ormqr(side blas.Side, trans blas.Transpose, a blas64.General, tau []float64, c blas64.General, work []float64, lwork int) {
+	lapack64.Dormqr(side, trans, c.Rows, c.Cols, a.Cols, a.Data, a.Stride, tau, c.Data, c.Stride, work, lwork)
+}
+
+// Trtrs solves a triangular system of the form A * X = B or A^T * X = B. Trtrs
+// returns whether A is singular. If A is singular, no solve is performed.
```
Comment (kortschak) [inline lapack64/lapack64.go]: ok != singular

**Comment by kortschak** [comment]
LGTM with minor comments.

File: cgo/lapack.go (hunk 2 above)
Comment (btracey) [inline cgo/lapack.go]: Thanks.

**Pull request merged**